File: hashmaps/find_words_that_can_be_formed_by_characters.py
class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        char_tally = Counter(chars)
        result = 0

        for word in words:
            word_tally = Counter(word)

            valid = True
            for char, count in word_tally.items():
                if char not in char_tally or char_tally[char] < count:
                    valid = False
                    break

            if valid:
                result += len(word)

        return result

File: hashmaps/find_words_that_can_be_formed_by_characters.rb
# @param {String[]} words
# @param {String} chars
# @return {Integer}
def count_characters(words, chars)
    char_tally = chars.chars.tally
    p char_tally

    result = 0

    words.each do |word|
        word_tally = word.chars.tally

        valid = true
        word_tally.each do |char, count|
            if !char_tally.key?(char) || char_tally[char] < count
                valid = false
                break
            end
        end

        result += word.size if valid
    end

    result
end